pkg/hypershift/defragcontroller/defragcontroller.go

```diff
@@ -0,0 +1,201 @@
+package defragcontroller
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/openshift/cluster-etcd-operator/pkg/etcdcli"
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/health"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+)
+
+const (
+	minDefragBytes          int64   = 100 * 1024 * 1024 // 100MB
+	minDefragWaitDuration           = 36 * time.Second
+	maxFragmentedPercentage float64 = 45
+	pollWaitDuration                = 2 * time.Second
+	pollTimeoutDuration             = 60 * time.Second
+	compactionInterval              = 10 * time.Minute
+)
+
+type HypershiftDefragController struct {
+	etcdClient etcdcli.EtcdClient
+
+	defragWaitDuration time.Duration
+}
+
+func NewHypershiftDefragController(livenessChecker *health.MultiAlivenessChecker, etcdClient etcdcli.EtcdClient, eventRecorder events.Recorder) factory.Controller {
+	c := &HypershiftDefragController{
+		etcdClient:         etcdClient,
+		defragWaitDuration: minDefragWaitDuration,
+	}
+	syncer := health.NewCheckingSyncWrapper(c.sync, 3*compactionInterval+1*time.Minute)
+	livenessChecker.Add("HypershiftDefragController", syncer)
+
+	// attempt to sync outside of etcd compaction interval to ensure maximum gain by defragmentation.
+	return factory.New().ResyncEvery(compactionInterval+1*time.Minute).
+		WithSync(syncer.Sync).
+		ToController("HypershiftDefragController", eventRecorder.WithComponentSuffix("hypershift-defrag-controller"))
+}
+
+func (c *HypershiftDefragController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	return c.runDefrag(ctx, syncCtx.Recorder())
+}
+
+func (c *HypershiftDefragController) runDefrag(ctx context.Context, recorder events.Recorder) error {
```

Comment:
no personal offense, but if you're already copying the whole controller - you might also copy it directly into your hypershift operator codebase. 

The PR at https://github.com/openshift/hypershift/pull/2869 looks pretty good already, doesn't it?

Reply:
non taken :)
We are trying to avoid dragging and maintaining so many new dependencies into our code base as you can see is the case with https://github.com/openshift/hypershift/pull/2869

The reason for copying the controller is to avoid refactoring the existing controller to work with Hypershift and end up breaking standalone.

Reply:
is there no other place you can build a dedicated binary for it, if dependencies are the concern? The etcdclient is not that bad to carry around... maybe GRPC is annoying to deal with.

@hasbro17 wdyt?

Reply:
> We are trying to avoid dragging and maintaining so many new dependencies into our code base as you can see is the case with https://github.com/openshift/hypershift/pull/2869

@muraee If we're copying the controller (with hypershift specific modifications that we're not quite aware of) then we're still maintaining it in two places. So we're not really reducing the maintenance burden by placing it in the cluster-etcd-operator vs the hypershift repo, since you still have to maintain it here.

Plus by sharing the underlying dependency instead of copying that too, there's a good chance we might unintentionally break hypershift's defrag cmd with future changes.

I'm not opposed to sharing dependencies where it makes sense but if we do this, then will we have alternative controllers/cmds for all the other pieces that hypershift might require from the CEO? e.g cert rotation or automated backups.

Reply:
btw, I've completely forgot about this. Ben wrote an upstream tool that also does that as a cmdline utility: https://github.com/ahrtr/etcd-defrag

there's plenty of config options to replicate our logic in CEO. That should also get you past the dependency issues.

Reply:
@tjungblu thanks for sharing the cmdline, we will look if its feasible for us to use it.

Reply:
@hasbro17 ideally we wouldn't have a seperate cmd for Hypershift, but the CEO would be Hypershift compatible and configurable enough, that we would just deploy it as it is.

But since that is not the case and we need to have the defragmentation feature before GA, the fasted solution was to create a separate cmd.

btw, we already have our own backup solution :)